pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -171,31 +201,52 @@ func createGatewayHTTPFilterChainOpts(
 	}
 
 	httpListeners := make([]*filterChainOpts, 0, len(servers))
-	for i, server := range servers {
-		routeCfg := buildGatewayInboundHTTPRouteConfig(env, nameToServiceMap, gatewayNames, server)
-		if routeCfg == nil {
-			log.Debugf("omitting HTTP listeners for port %d filter chain %d due to no routes", server.Port, i)
-			continue
-		}
-		// if https redirect is set, we need to enable requireTls field in all the virtual hosts
-		if server.Tls != nil && server.Tls.HttpsRedirect {
-			for i := range routeCfg.VirtualHosts {
-				// TODO: should this be set to ALL ?
-				routeCfg.VirtualHosts[i].RequireTls = route.VirtualHost_EXTERNAL_ONLY
-			}
-		}
+	// Are we processing plaintext servers or TLS servers?
+	// If plain text, we have to combine all servers into a single listener
```

Comment:
I think there is a small problem here: you seem to assume SNI will be used and for TLS we can have on virtual host, while plain text will have all hosts.

I think the most common case ( on real world hosts ) is multi-SAN certs - where the plain text and TLS listeners have the same set of virtual hosts.

Reply:
I agree, but the code handles that as well. you will end up having one RDS route for https with all the virtual hosts (server names in https port) and another RDS for http, with all the virtual hosts (server names in the http server). I feel that trying to create a single RDS for two servers by matching the list of hosts they expose - is an optimization that might quickly become useless the moment the set of hosts in the two servers start differing. At that point, we would still need code to handle the scenario that I am handling today. Put another way, the code as it is today, will result in two RDS calls (one for http, anbd one for the multi-san https), while what you are suggesting will require more code to collapse this into a single RDS call, and yet require the logic above to handle cases where two servers differ

Reply:
Is there any legitimate case where the 2 would differ ? How will it be handled by the 'gradual adoption' stuff ? Seems complex and dangerous. I don't mind your implementation that much, but the use case.

BTW, SNI in general doesn't mean single-SAN, it is possible to have multiple multi-SAN certificates, in addition to the the default cert which usually is multi-SAN.
I believe envoy handles this correctly but we don't generate code for it - i.e. you can have a list
of hostnames on the same chain, matching multiple SNI. 

But my primary concern remains RDS for plain text and the 'default/no-SNI' -  if you think they need different names I'm fine, as long as current code generates the same list of hosts for both. We can debate the use case for having different hosts and routes in separate PR :-)

I wouldn't mind if the SNI case would be handled in separate PR as well - this is gateway specific and it may be better to move on with RDS testing for in-cluster plus the gatway plain text/default tls.

Reply:
> BTW, SNI in general doesn't mean single-SAN, it is possible to have multiple multi-SAN certificates, in addition to the the default cert which usually is multi-SAN.
I believe envoy handles this correctly but we don't generate code for it - i.e. you can have a list
of hostnames on the same chain, matching multiple SNI.

We do. All hosts in the server go into a single filter chain match.

Reply:
> But my primary concern remains RDS for plain text and the 'default/no-SNI' - if you think they need different names I'm fine, as long as current code generates the same list of hosts for both. We can debate the use case for having different hosts and routes in separate PR :-)

so look at it this way: The gateway API spec is like this:

```yaml
- server:
   port: 80
   hosts:
   - foo.com
   - bar.com...
- server:
     port: 443
   hosts:
   - foo.com
   - bar.com
   - scooby.com
   tls:
      keys: ...
      ca: ...
```

What I am saying is, in order to have the same RDS route config (e.g., http.80) shared between servers on port 80 and port 443, we have to compare the list of hosts in both servers and ensure that they are the same. Only then can we use the same RDS route config for both servers. 99.99% of the time, it is going to be the same as you said. But, the moment someone adds an extra host on HTTP but not on HTTPS, we have to handle the scenario by having two different RDS route configs (one for each server). This is not possible just by RDS, because the listeners on port 80 and 443 are already pointing to the shared RDS (e.g., http.80). So, you would now have to reload the 443 listener to supply it with a new RDS name in its http connection manager and then do the RDS call. All this logic will have to be added or we have to tell people that the HTTP servers in the gateway can only have same set of hosts (which doesn't make sense).

Now, multiply this by N servers (gateway allows any number of servers,  unlike kube ingress), and any number of hosts, etc. The code to detect and squash multiple servers into a shared RDS route config will be equally big, plus the code to cause LDS reload in case one of the servers starts having different sets of servers, and a whole host more.

And in terms of SNI, we certainly do support single large cert serving multiple hosts, because to us its just a single cert and a filter chain match that has one or more hosts configured in teh SNI match section.